lib/Sema/TypeCheckPattern.cpp

```diff
@@ -1282,20 +1282,27 @@ Pattern *TypeChecker::coercePatternToType(ContextualPattern pattern,
     }
 
     // case nil is equivalent to .none when switching on Optionals.
-    if (type->getOptionalObjectType()) {
-      auto EP = cast<ExprPattern>(P);
-      if (auto *NLE = dyn_cast<NilLiteralExpr>(EP->getSubExpr())) {
+    if (auto *NLE = dyn_cast<NilLiteralExpr>(EP->getSubExpr())) {
+      if (type->getOptionalObjectType()) {
         auto *NoneEnumElement = Context.getOptionalNoneDecl();
         auto *BaseTE = TypeExpr::createImplicit(type, Context);
         P = new (Context) EnumElementPattern(
             BaseTE, NLE->getLoc(), DeclNameLoc(NLE->getLoc()),
             NoneEnumElement->createNameRef(), NoneEnumElement, nullptr);
         return TypeChecker::coercePatternToType(
             pattern.forSubPattern(P, /*retainTopLevel=*/true), type, options);
+      } else {
+        // ...but for non-optional types it can never match! Diagnose it.
```

Comment:
@CodaFi Will this trip up when switching over values of non-`Optional` types that conform to `ExpressibleByNilLiteral`?

Reply:
It will not. I made sure of this locally - in fact it does the worst possible thing and up-converts the source value into an optional then applies `~=` rather than applying the `ExpressibleByNilLiteral` conversion. I should have committed that into the regression test suite though. I'll get that with a quick follow-on.

Reply:
It will not. I made sure of this locally - in fact it does the worst possible thing and up-converts the source value into an optional then applies `~=` rather than applying the `ExpressibleByNilLiteral` conversion. I should have committed that into the regression test suite though. I'll get that with a quick follow-on.